### PR TITLE
[InstCombine] Reduce range of ctpop for non zero argument

### PR DESCRIPTION
**Pull request by andjo403**
Part of solution for #95255

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp [hunk 1]
```diff
@@ -694,8 +694,13 @@ static Instruction *foldCtpop(IntrinsicInst &II, InstCombinerImpl &IC) {
   // Add range attribute since known bits can't completely reflect what we know.
   if (BitWidth != 1 && !II.hasRetAttr(Attribute::Range) &&
       !II.getMetadata(LLVMContext::MD_range)) {
-    ConstantRange Range(APInt(BitWidth, Known.countMinPopulation()),
-                        APInt(BitWidth, Known.countMaxPopulation() + 1));
+    unsigned KnownMinPop = Known.countMinPopulation();
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:697]: There is a phase ordering problem. We may get a more precise result later. See my previous comment https://github.com/llvm/llvm-project/pull/100870#issuecomment-2254213148.

Comment (andjo403) [reply llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:697]: did not think that far, forgets that the checks id done multiple times.
but then it fells like the draft that you have is better then this solution so we do not need to execute isKnownNonZero if not needed.

Comment (andjo403) [reply llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp]: added the test and when for a combination of our solutions that reduced number of calls to intersectWith also used full set range if old is missing as it is faster in intersectWith

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!

**Comment by goldsteinn** [comment]
LGTM

**Review by nikic — APPROVED** [review]
LGTM

**Comment by pranavk** [comment]
This seems to cause miscompiles for one of our Arm targets. Here's the reproducer:
https://godbolt.org/z/689eKhcnh

The problem seems to be the range attribute that doesn't contain "0" even though smax could return 0 in this case. This then combines with https://github.com/llvm/llvm-project/pull/111998/files which causes incorrect value of b on Arm machines.

**Comment by andjo403** [comment]
this pass of simplifycfg https://godbolt.org/z/PhqdMcWG3 looks strange to me starts to call ctpop for zero but do not drop the range attribute

**Comment by dtcxzyw** [comment]
> this pass of simplifycfg https://godbolt.org/z/PhqdMcWG3 looks strange to me starts to call ctpop for zero but do not drop the range attribute

It is correct. Even if `ctpop` returns a poison value due to range violation, `%6 = select i1 %3, i1 %5, i1 false` still returns false.

**Comment by dtcxzyw** [comment]
> This seems to cause miscompiles for one of our Arm targets. Here's the reproducer: https://godbolt.org/z/689eKhcnh
> 
> The problem seems to be the range attribute that doesn't contain "0" even though smax could return 0 in this case. This then combines with https://github.com/llvm/llvm-project/pull/111998/files which causes incorrect value of b on Arm machines.

The result looks OK? https://godbolt.org/z/T5E1oYvse
`bit_ceil(0) = 1, is_power_of_two(0) = 0`

**Comment by pranavk** [comment]
In the original repro, I am using absl and c++17. https://godbolt.org/z/35c9xfn6h but I can't make executor run this because it can't find absl headers.

**Comment by pranavk** [comment]
When I try to execute it on an Arm machine, I get b = INT_MAX.

**Comment by dtcxzyw** [comment]
> In the original repro, I am using absl and c++17. https://godbolt.org/z/35c9xfn6h but I can't make executor run this because it can't find absl libraries.

```
define dso_local noundef range(i32 1, -2147483647) i32 @bucket(int)(i32 noundef %0) local_unnamed_addr #4 { %0 -> 0
  %2 = tail call i32 @llvm.smax.i32(i32 %0, i32 0) -> 0
  %3 = icmp sgt i32 %0, 0 -> false
  %4 = tail call range(i32 1, 32) i32 @llvm.ctpop.i32(i32 %2) -> poison
  %5 = icmp samesign ult i32 %4, 2 -> poison
  %6 = select i1 %3, i1 %5, i1 false -> false
  %7 = tail call range(i32 0, 33) i32 @llvm.ctlz.i32(i32 %2, i1 true) -> poison
  %8 = add nsw i32 %2, -1 -> -1
  %9 = tail call range(i32 0, 33) i32 @llvm.ctlz.i32(i32 %8, i1 false) -> 0
  %10 = select i1 %6, i32 %7, i32 %9 -> 0
  %11 = sub nsw i32 0, %10 -> 0
  %12 = and i32 %11, 31 -> 0
  %13 = shl nuw i32 1, %12 -> 1
  ret i32 %13 -> 1
}
```

Looks like it is a codegen bug in ARM backend.

**Comment by dtcxzyw** [comment]
I will file an issue later. It seems to be a SDAG bug: https://godbolt.org/z/W6cdshn7q

**Comment by dtcxzyw** [comment]
> this pass of simplifycfg https://godbolt.org/z/PhqdMcWG3 looks strange to me starts to call ctpop for zero but do not drop the range attribute

You are correct.

**Pull request merged**